daemon/emer-aggregate-tally.c

```diff
@@ -0,0 +1,391 @@
+                                                    date);
+
+  variant = load_aggregate_timer_from_path (tally_path);
+
+  if (variant)
+    {
+      g_variant_get_child (variant,
+                           3,
+                           "u",
+                           &previous_counter);
+    }
+
+  return previous_counter;
+}
+
+static void
+emer_aggregate_tally_finalize (GObject *object)
+{
+  EmerAggregateTally *self = (EmerAggregateTally *)object;
+
+  g_clear_pointer (&self->persistent_cache_directory, g_free);
+
+  G_OBJECT_CLASS (emer_aggregate_tally_parent_class)->finalize (object);
+}
+
+static void
+emer_aggregate_tally_set_property (GObject      *object,
+                                   guint         prop_id,
+                                   const GValue *value,
+                                   GParamSpec   *pspec)
+{
+  EmerAggregateTally *self = EMER_AGGREGATE_TALLY (object);
+
+  switch (prop_id)
+    {
+    case PROP_PERSISTENT_CACHE_DIRECTORY:
+      g_assert (self->persistent_cache_directory == NULL);
+      self->persistent_cache_directory = g_value_dup_string (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+emer_aggregate_tally_class_init (EmerAggregateTallyClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = emer_aggregate_tally_finalize;
+  object_class->set_property = emer_aggregate_tally_set_property;
+
+  /*
+   * EmerDaemon:persistent-cache-directory:
+   *
+   * A directory for temporarily storing events until they are uploaded to the
+   * metrics servers.
+   */
+  properties[PROP_PERSISTENT_CACHE_DIRECTORY] =
+    g_param_spec_string ("persistent-cache-directory",
+                         "Persistent cache directory",
+                         "The directory in which to temporarily store events "
+                         "locally",
+                         NULL,
+                         G_PARAM_CONSTRUCT_ONLY | G_PARAM_WRITABLE |
+                         G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class, N_PROPS, properties);
+}
+
+static void
+emer_aggregate_tally_init (EmerAggregateTally *self)
+{
+}
+
+EmerAggregateTally *
+emer_aggregate_tally_new (const char *persistent_cache_directory)
+{
+  return g_object_new (EMER_TYPE_AGGREGATE_TALLY,
+                       "persistent-cache-directory", persistent_cache_directory,
+                       NULL);
+}
+
+void
+emer_aggregate_tally_store_event (EmerAggregateTally *self,
+                                  guint32             unix_user_id,
+                                  GVariant           *event_id,
+                                  GVariant           *aggregate_key,
+                                  GVariant           *payload,
+                                  guint32             counter,
+                                  const char         *date,
```

Comment:
I think it would be preferable to use `GDate *` in the API rather than this "stringly-typed" approach.

Reply:
Oh… it's a string because it can be either yyyy-mm-dd or yyyy-mm. I still think we could stand to enforce one or the other in the API.

Reply:
> I still think we could stand to enforce one or the other in the API.

Or define a helper type which enforces one or the other, and has some utility functions for converting to/from strings/`GDate`s/`GDateTime`s